### PR TITLE
changes to assemble the complete subject hierarchy from the cache layer

### PR DESCRIPTION
**Pull request by shriramshankar**
We need to merge PR  #262  and #264 before this. I have checked in @pallavi2209's redisTestUtil.js file too for the tests to pass.  I have also done some refracting for the cache layer and the DB layer to be able to use the same set of filters.

### CODE REVIEW

File: tests/cache/models/subjects/getHierarchy.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+      .set('Authorization', token)
+      .expect(constants.httpStatus.OK)
+      .expect((res) => {
+        expect(res.body.children[0].children[0].samples).to.be
+          .an('array');
+        console.log(res.body);
+        expect(res.body.children[0].children[0].samples).to.have.lengthOf(1);
+        expect(res.body.children[0].children[0].samples[0].aspect.rank).
+          to.equal(10);
+      })
```
Comment (shriramshankar) [inline tests/cache/models/subjects/getHierarchy.js]: need to talk about this. returns '10' instead of 10.

File: cache/models/subject.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+/**
+ * Given a subject with its samples, aspect, aspectTags and sampleStatus filters
+ * are applied to the samples and the filtered samples are attached back to the
+ * subject
+ * @param  {Object} subj - Subject object on which the filters are applied
+ * @returns {Object} - filtered subject
+ */
+function filterSamples(subj) {
+  const filteredSamples = [];
```
Comment (shriramshankar) [inline cache/models/subject.js]: @iamigo @pallavi2209  I moved the aspect filters along with the other filters. Applying just the aspect includes filter while getting it from redis (using "hmget aspectnames")  will work but adding the excludes filter will nest the promises and break the promise chain. Any thoughts about this?

File: cache/models/subject.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  return redisClient.smembersAsync(subjectKey).then((aspectNames) => {
+    // console.log(1);
+    const sampleAspectPromise = [];
+    aspectNames.forEach((aspect) => {
+      const sampleName = res.absolutePath + '|' + aspect;
+      sampleAspectPromise.push(getSampleFromRedis(sampleName));
+      sampleAspectPromise.push(getAspectFromRedis(aspect));
+    });
+
+    return Promise.all(sampleAspectPromise);
```
Comment (pallavi2209) [inline cache/models/subject.js]: We can do this in one redis call using commands array as we discussed.

**Comment by shriramshankar** [comment]
@iamigo or @pallavi2209 Can one of you merge this PR? I think I skipped a patch that I should not have while rebasing and my workspace is kind of messed up now. I will create another PR with the changes in the comments.

**Comment by pallavi2209** [comment]
I am fine merging this if @iamigo approves.

**Pull request merged**